### PR TITLE
[Serve] add dependency management

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Uses environment variable overrides for actors (https://github.com/ray-project/ray/pull/11619) to allow simultaneously serving backends with conflicting dependencies, using `conda` environments.  The environments must be preinstalled on all nodes.

Usage example:
```python
client.create_backend("tf1", tf_version, env=CondaEnv("ray-tf1"))
```

## Related issue number

Comments on an earlier version of this PR can be found at https://github.com/architkulkarni/ray/pull/1.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Tested manually
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — APPROVED** [review]
LGTM

**Comment by edoakes** [comment]
@architkulkarni looks like some of the CI jobs failed mysteriously. Restarting, should be good to merge after that.

**Comment by architkulkarni** [comment]
I just remembered that CI runs all the doc examples as tests.  This is why the CI build is failing, because the conda envs in my doc example aren't set up.  Is there a way to make CI skip this example?

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -8,11 +8,14 @@
...
 from ray.serve.utils import (block_until_http_ready, format_actor_name,
-                             get_random_letters, logger, get_node_id_for_actor)
+                             get_random_letters, logger, get_node_id_for_actor,
+                             get_conda_env_dir)
 from ray.serve.exceptions import RayServeException
 from ray.serve.config import BackendConfig, ReplicaConfig, BackendMetadata
+from ray.serve.env import CondaEnv
 from ray.actor import ActorHandle
 from typing import Any, Callable, Dict, List, Optional, Type, Union
+import os
```
Comment (simon-mo) [inline python/ray/serve/api.py:18]: can you move the import to line 4? stdlib should be groupped together, see https://google.github.io/styleguide/pyguide.html#313-imports-formatting

File: python/ray/serve/tests/test_util.py [hunk 2]
```diff
@@ -1,12 +1,13 @@
 import asyncio
 import json
 from copy import deepcopy
-
+import os
```
Comment (simon-mo) [inline python/ray/serve/tests/test_util.py:4]: move this import to first group stdlib

File: python/ray/serve/utils.py [hunk 3]
```diff
@@ -179,6 +179,22 @@ def format_actor_name(actor_name, controller_name=None, *modifiers):
...
+    if os.environ.get("CONDA_DEFAULT_ENV") == "base":
+        # Caller is running in base conda env.
+        # Not recommended by conda, but we can still try to support it.
+        env_dir = os.path.join(conda_prefix, "envs", env_name)
+    else:
+        env_dir = os.path.join(
+            os.path.split(os.environ.get("CONDA_PREFIX"))[0], env_name)
+    if not os.path.isdir(env_dir):
+        raise ValueError("conda env " + env_name + " not found.")
+    return env_dir
```
Comment (simon-mo) [inline python/ray/serve/utils.py:195]:
```suggestion
def get_conda_env_dir(env_name):
    """Given a environment name like `tf1`, find and validate the corresponding conda directory"""
    conda_prefix = os.environ.get("CONDA_PREFIX")
    if conda_prefix is None:
        raise ValueError("Serve cannot find environment variables installed by conda. Are you sure you are in a conda env?")
    
    # There are two case:
    # 1. We are in conda base env: CONDA_DEFAULT_ENV=base and CONDA_PREFIX=$HOME/anaconda3
    # 2. We are in user created conda env: CONDA_DEFAULT_ENV=$env_name and CONDA_PREFIX=$HOME/anaconda3/envs/$env_name
    if os.environ.get("CONDA_DEFAULT_ENV") == "base":
        # Caller is running in base conda env.
        # Not recommended by conda, but we can still try to support it.
        env_dir = os.path.join(conda_prefix, "envs", env_name)
    else:
        # Now `conda_prefix` should be something like $HOME/anaconda3/envs/$env_name
        # We want to strip the $env_name component
        conda_envs_dir = os.path.split(conda_prefix)[0]
        env_dir = os.path.join(conda_envs_dir, env_name)
    if not os.path.isdir(env_dir):
        raise ValueError("conda env " + env_name + " not found in conda envs directory. Run `conda env list` to verify the name is correct.")
    return env_dir
```

Comment (simon-mo) [reply python/ray/serve/utils.py:195]: Some suggestions to make the code a little bit more readable and maintainable. You might need to fix some typos and lint.

File: python/ray/serve/api.py (hunk 1 above)
Comment (architkulkarni) [reply python/ray/serve/api.py:18]: Sure thing, thanks for the reference!

**Pull request merged**